File: uniffi_bindgen/src/bindings/kotlin/templates/TraitObjectImpl.kt
{%- let handle_map = "uniffiTraitHandleMap{name}"|format %}
internal val {{ handle_map }} = UniFfiHandleMap<{{ interface_name }}>()

@Structure.FieldOrder({% for meth in obj.methods() %}"{{ meth.name() }}", {% endfor %}"uniffi_free")
class {{ name|trait_vtable_class }} : Structure() {
    // For each method, define an interface for that method and a nested class that implements it.
    // It would be nice to use a Kotlin `object`, but that doesn't work with JNA.
    {%- for meth in obj.methods() %}
    internal interface {{ meth.name()|class_name }}Interface : com.sun.jna.Callback {
        fun invoke({{ meth|trait_vtable_method_args }})
    }

    internal class {{ meth.name()|class_name }}Impl : {{ meth.name()|class_name }}Interface {
        override fun invoke({{ meth|trait_vtable_method_args }}) {
            val uniffiObj = {{ handle_map }}.get(handle) ?: throw NullPointerException("Invalid handle passed to {{ name }}.{{ meth.name() }}")
            val makeCall = { ->
                uniffiObj.{{ meth.name()|fn_name() }}(
                    {%- for arg in meth.arguments() %}
                    {{ arg|lift_fn }}({{ arg.name() }}),
                    {%- endfor %}
                )
            }

            {%- match meth.return_type() %}
            {%- when Some(return_type) %}
            val writeReturn = { value: {{ return_type|type_name }} -> uniffiOutReturn.setValue({{ return_type|lower_fn }}(value)) }
            {%- when None %}
            val writeReturn = { _: Unit -> Unit }
            {%- endmatch %}

            {%- match meth.throws_type() %}
            {%- when None %}
            uniffiTraitInterfaceCall(uniffiOutCallStatus, makeCall, writeReturn)
            {%- when Some(error_type) %}
            uniffiTraitInterfaceCallWithError(
                uniffiOutCallStatus,
                makeCall,
                writeReturn,
                { e: {{error_type|error_type_name }} -> {{ error_type|lower_fn }}(e) }
            )
            {%- endmatch %}
        }
    }

    {%- endfor %}

    internal interface UniffiFreeInterface : com.sun.jna.Callback {
        fun invoke(handle: USize);
    }

    internal class UniffiFreeImpl : UniffiFreeInterface {
        override fun invoke(handle: USize) {
            {{ handle_map }}.remove(handle)
        }
    }

    {%- for meth in obj.methods() %}
    @JvmField internal var {{ meth.name() }} : {{ meth.name()|class_name }}Interface = {{ meth.name()|class_name }}Impl()
    {%- endfor %}
    @JvmField internal var uniffi_free : UniffiFreeInterface = UniffiFreeImpl()

    internal fun initialize(lib: _UniFFILib) {
        lib.{{ obj.ffi_init_trait_callback().name() }}(this)
    }
}

val {{ name|trait_vtable_obj }} : {{ name|trait_vtable_class }} = {{ name|trait_vtable_class }}()

File: uniffi_bindgen/src/bindings/python/templates/TraitObjectImpl.py
# Put all the bits inside a class to keep the top-level namespace clean
{%- let trait_impl="UniffiTraitImpl{name}"|format %}
class {{ trait_impl }}:
    _UniffiPointerManager = _UniffiPointerManager()
    {%- for meth in obj.methods() %}

    @{{ meth.ffi_func()|ctypes_ffi_prototype }}
    def {{ meth.name() }}(
            uniffi_handle,
            {%- for arg in meth.arguments() %}
            {{ arg.name() }},
            {%- endfor %}
            return_ptr,
            uniffi_call_status_ptr,
        ):
        uniffi_obj = {{ trait_impl }}._UniffiPointerManager.lookup(uniffi_handle)
        def make_call():
            args = ({% for arg in meth.arguments() %}{{ arg|lift_fn }}({{ arg.name() }}), {% endfor %})
            method = uniffi_obj.{{ meth.name() }}
            return method(*args)
        {%- match meth.return_type() %}
        {%- when Some(return_type) %}
        def write_return_value(v):
            return_ptr[0] = {{ return_type|lower_fn }}(v)
        {%- when None %}
        write_return_value = lambda v: None
        {%- endmatch %}

        {%- match meth.throws_type() %}
        {%- when None %}
        _uniffi_trait_interface_call(
                uniffi_call_status_ptr.contents,
                make_call,
                write_return_value,
        )
        {%- when Some(error) %}
        _uniffi_trait_interface_call_with_error(
                uniffi_call_status_ptr.contents,
                make_call,
                write_return_value,
                {{ error|type_name }},
                {{ error|lower_fn }},
        )
        {%- endmatch %}
    {%- endfor %}

    @ctypes.CFUNCTYPE(None, ctypes.c_void_p)
    def uniffi_free(uniffi_handle):
        {{ trait_impl }}._UniffiPointerManager.release_pointer(uniffi_handle)

    class UniffiVtable(ctypes.Structure):
        _fields_ = [
            {%- for meth in obj.methods() %}
            ("{{ meth.name() }}", {{ meth.ffi_func()|ctypes_ffi_prototype }}),
            {%- endfor %}
            ("uniffi_free", ctypes.CFUNCTYPE(None, ctypes.c_void_p)),
    ]
    uniffi_vtable = UniffiVtable(
        {%- for meth in obj.methods() %}
        {{ meth.name() }},
        {%- endfor %}
        uniffi_free
    )
    _UniffiLib.{{ obj.ffi_init_trait_callback().name() }}(ctypes.byref(uniffi_vtable))
